src/lib/parallel_scan/parallel_scan.ml

```diff
@@ -2,925 +2,1533 @@ open Core_kernel
+
+  let _fold_until : type merge_t base_t accum final.
+         f_merge:(accum -> merge_t -> (accum, final) Continue_or_stop.t)
+      -> f_base:(accum -> base_t -> (accum, final) Continue_or_stop.t)
+      -> init:accum
+      -> finish:(accum -> final)
+      -> (merge_t, base_t) t
+      -> final =
+   fun ~f_merge ~f_base ~init ~finish t ->
+    Foldable_ident.fold_depth_until
+      ~f_merge:(fun _ -> f_merge)
+      ~f_base ~init ~finish t
+
+  (*
+    result -> final proof
+    f_merge, f_base are to update the nodes with new jobs and mark old jobs as "Done"*)
+  let rec update_split : type merge_t base_t data weight result.
+         f_merge:(data -> int -> merge_t -> merge_t * result option)
+      -> f_base:(data -> base_t -> base_t)
+      -> weight_merge:(merge_t -> weight * weight)
+      -> jobs:data
+      -> update_level:int
+      -> jobs_split:(weight * weight -> data -> data * data)
+      -> (merge_t, base_t) t
+      -> (merge_t, base_t) t * result option =
+   fun ~f_merge ~f_base ~weight_merge ~jobs ~update_level ~jobs_split t ->
+    match t with
+    | Leaf d ->
+        (Leaf (f_base jobs d), None)
+    | Node {depth; value; sub_tree} ->
+        let weight_left_subtree, weight_right_subtree = weight_merge value in
+        (*update the jobs at the current level*)
+        let value', scan_result = f_merge jobs depth value in
+        (*get the updated subtree*)
+        let sub, _ =
+          if update_level = depth then (sub_tree, None)
+          else
+            (*split the jobs for the next level*)
+            let new_jobs_list =
+              jobs_split (weight_left_subtree, weight_right_subtree) jobs
+            in
+            update_split
+              ~f_merge:(fun (b, b') i (x, y) ->
+                let left = f_merge b i x in
+                let right = f_merge b' i y in
+                ((fst left, fst right), Option.both (snd left) (snd right)) )
+              ~f_base:(fun (b, b') (x, x') -> (f_base b x, f_base b' x'))
+              ~weight_merge:(fun (a, b) -> (weight_merge a, weight_merge b))
+              ~update_level
+              ~jobs_split:(fun (x, y) (a, b) ->
+                (jobs_split x a, jobs_split y b) )
+              ~jobs:new_jobs_list sub_tree
+        in
+        (Node {depth; value= value'; sub_tree= sub}, scan_result)
+
+  let rec update_accumulate : type merge_t base_t data.
+         f_merge:(data * data -> merge_t -> merge_t * data)
+      -> f_base:(base_t -> base_t * data)
+      -> (merge_t, base_t) t
+      -> (merge_t, base_t) t * data =
+   fun ~f_merge ~f_base t ->
+    match t with
+    | Leaf d ->
+        let new_base, count_list = f_base d in
+        (Leaf new_base, count_list)
+    | Node {depth; value; sub_tree} ->
+        (*get the updated subtree*)
+        let sub, counts =
+          update_accumulate
+            ~f_merge:(fun b (x, y) ->
+              let b1, b2 = b in
+              let left, count1 = f_merge b1 x in
```

Comment:
Maybe define a function 

transpose ((x1, y1), (x2, y2)) = ((x1, x2), (y1, y2)) 

and use that